File: 核函數主成分分析/kernel_pca.py
#核函數主成分分析(kernel PCA)
#用於線性不可分數據
#套件
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import KernelPCA
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from matplotlib.colors import ListedColormap

#read
data = pd.read_csv('Social_Network_Ads.csv')
x = data.iloc[:, [2, 3]].values
y = data.iloc[:, 4].values

#train、test
x_train, x_test, y_train, y_test = train_test_split(
        x, y, test_size = 0.2)

#feature scaling
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test) #上面fit過了

#apply kernel-PCA 
kpca = KernelPCA(n_components = 2) #n_components為提取的自變數數量
x_train = kpca.fit_transform(x_train)
x_test = kpca.transform(x_test) #上面fit過了

#Logistic regression
model = LogisticRegression()
model.fit(x_train, y_train)

#predict
y_pred = model.predict(x_test)

#confusion matrix
cm = confusion_matrix(y_test, y_pred)

#visual(train_set)
x_set, y_set = x_train, y_train
#網格(上下左右留白一個單位)
x1, x2 = np.meshgrid(np.arange(x_set[:, 0].min() - 1, x_set[:, 0].max() + 1,
                               step = 0.01),
                    np.arange(x_set[:, 1].min() - 1, x_set[:, 1].max() + 1,
                              step = 0.01))
#分界線
plt.contourf(x1, x2, model.predict(np.array([x1.ravel(), x2.ravel()]).T).reshape(
        x1.shape), alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())
#將點上色
#迴圈
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('orange', 'blue'))(i), label = j, s = 5)
plt.title('Logistic Regression')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.legend()
plt.show()

#test_set
x_set, y_set = x_test, y_test
#網格(上下左右留白一個單位)
x1, x2 = np.meshgrid(np.arange(x_set[:, 0].min() - 1, x_set[:, 0].max() + 1,
                               step = 0.01),
                    np.arange(x_set[:, 1].min() - 1, x_set[:, 1].max() + 1,
                              step = 0.01))
#分界線
plt.contourf(x1, x2, model.predict(np.array([x1.ravel(), x2.ravel()]).T).reshape(
        x1.shape), alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(x1.min(), x1.max())
plt.ylim(x2.min(), x2.max())
#將點上色
#迴圈
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('orange', 'blue'))(i), label = j, s = 5)
plt.title('Logistic Regression')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.legend()
plt.show()

File: 核函數主成分分析/kernel_pca.r
#核函數主成分分析(kernel PCA)

#read
data = read.csv('Social_Network_Ads.csv')
data = data[, 3:5]

#train、test
library(caTools)
split = sample.split(data$Purchased, SplitRatio = 0.75)
train_set = subset(data, split == TRUE)
test_set = subset(data, split == FALSE)

#feature scaling(col:1,2)
train_set[, 1:2] = scale(train_set[, 1:2])
test_set[, 1:2] = scale(test_set[, 1:2])

#apply kernel-PCA
#features為提取自變數個數
library(kernlab)
kpca = kpca(~., data = train_set[, -3], kernel = 'rbfdot', features = 2)
#predict結果為矩陣，必須轉換成dataframe
train_set_pca = as.data.frame(predict(kpca, train_set))
#增加應變數
train_set_pca$Purchased = train_set$Purchased
test_set_pca = as.data.frame(predict(kpca, test_set))
test_set_pca$Purchased = test_set$Purchased

#build Logistic regresssion
model = glm(Purchased ~ ., family = binomial, data = train_set_pca)

#predict
prob_pred = predict(model, type = 'response', newdata = test_set_pca[, -3])
#將機率轉換成0或1
y_pred = ifelse(prob_pred > 0.5, 1, 0)

#confusion matrix
cm = table(test_set_pca[, 3], y_pred)

#visual
#train_set
library(ElemStatLearn)
set = train_set_pca
#網格
x1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.0075)
x2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.0075)
grid_set = expand.grid(x1, x2)
colnames(grid_set) = c('V1', 'V2')
#predict grid_set
prob_set = predict(model, type = 'response', grid_set)
#將prob_set改成1或0
y_grid = ifelse(prob_set > 0.5, 1, 0)
#畫點
plot(set[, -3],
     main = 'train_set', xlab = 'V1', ylab = 'V2',
     xlim = range(x1), ylim = range(x2))
#分界線
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)), add = TRUE)
#將分界區域上色
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
#將點上色
points(set[, c(1,2)], pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

#test_set
set = test_set_pca
#網格
x1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.0075)
x2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.0075)
grid_set = expand.grid(x1, x2)
colnames(grid_set) = c('V1', 'V2')
#predict grid_set
prob_set = predict(model, type = 'response', grid_set)
#將prob_set改成1或0
y_grid = ifelse(prob_set > 0.5, 1, 0)
#畫點
plot(set[, -3],
     main = 'train_set', xlab = 'V1', ylab = 'V2',
     xlim = range(x1), ylim = range(x2))
#分界線
contour(x1, x2, matrix(as.numeric(y_grid), length(x1), length(x2)), add = TRUE)
#將分界區域上色
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
#將點上色
points(set[, c(1,2)], pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

